MAGN-8959 As a user, I want to SEE watch nodes in a frozen state so that I can tell that they are not being executed

ramramps opened the pull request:
## Purpose

The purpose of this PR is to freeze the watch node and the watch node does not get evaluated until it is unfreeze. 

This PR refers to the below tasks 
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8959
http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9096
 
Here the task is to freeze the execution of watch node. 
Refer to  this PR  [https://github.com/DynamoDS/Dynamo/pull/5669] for the watch node UI changes.

In this PR, the following tasks are accompolished
 1.When a node is Freeze, the graph does not execute. Instead, the node and all its dependencies are deleted from AST. So, all the frozen nodes never participates in any of the execution.

 2.When a node is unfreeze, all its downstream nodes are set to modified, and the graph will be executed. 

3.Any time, when an input node is in freeze state, the value from the preview run is maintained. 

4.If the value of a freeze node is changed, the value will be changed in UI. The value will not get propagated to VM until the node is made unfreeze.

## Design Document
 https://docs.google.com/a/adsk-oss.com/presentation/d/1MUhSzoDo-QFPysDv1E6gRslpq2pnb6eoPjKSsJN6g-o/edit?usp=sharing  

## Declaration
Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

## Reviewers
@pboyer 

FYI @mjkkirschner

ramramps commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -602,6 +602,37 @@ public void NodeDeleted(NodeModel node)
+        /// deleted from AST
+        /// </summary>
+        /// <param name="node">The node.</param>
+        public void DeleteFrozenNodesFromAST(NodeModel node)
```
All the down stream nodes are deleted from AST.

ramramps commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -633,8 +633,20 @@ public bool IsFrozen
+                RaisePropertyChanged("IsFrozen");      
+                //If the node is Unfreezed then Mark all the downstream nodes as
+                // modified. This is essential recompiling the AST.
+                if (!value)
```
If the freeze is false, then  set the downstream nodes to be modified. this helps when compiling the graph.

ramramps commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -1713,7 +1747,7 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
                 RaisePropertyChanged("IsVisible");
                 RaisePropertyChanged("IsUpstreamVisible");
-
+               
```
Oops. sorry for the white space.

ramramps commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -740,7 +764,7 @@ public void SetNickNameFromAttribute()
         public virtual void OnNodeModified(bool forceExecute = false)
         {
-            if (!RaisesModificationEvents)
+            if (!RaisesModificationEvents || IsFrozen)
```
If a node is frozen, then that node does not participate in the execution

pboyer replied:
Are you sure this won't have unintended consequences?

pboyer replied:
We should have a test for this behavior, I think.

ramramps replied:
We have NodeExecutionTest class. If a node is frozen and modified, they will not execute and the result remains the same.All the tests  in NodeExeuctionTest class, tests this functionality.

ramramps commented on src/DynamoCoreWpf/ViewModels/Core/NodeViewModel.cs:
```diff
@@ -636,7 +636,11 @@ void logic_PropertyChanged(object sender, PropertyChangedEventArgs e)
-                    break;                             
+                    break;   
+                case "IsFrozen":
+                    RaisePropertyChanged("IsFrozenExplicitly");
```
This is required here to reflect the changes on UI

pboyer commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -602,6 +602,37 @@ public void NodeDeleted(NodeModel node)
+
+        private static void GetDownstreamNodes(NodeModel node, ICollection<NodeModel> gathered)
+        {
+            if (gathered.Contains(node)) // Considered this node before, bail.pu
```
"bail.pu"?

pboyer replied:
Again, a `HashSet` would be more appropriate.

pboyer commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -602,6 +602,37 @@ public void NodeDeleted(NodeModel node)
+            gathered.Add(node);
+
+            var sets = node.OutputNodes.Values;
+            var outputNodes = sets.SelectMany(set => set.Select(t => t.Item2)).Distinct();
```
A `HashSet` would do this more efficiently.

pboyer commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -668,6 +681,18 @@ private bool CheckIfAnyUpstreamNodeIsFrozen(NodeModel node, ref bool ret)
         }
+
+        private void MarkDownStreamNodesAsModified(NodeModel node)
+        {                         
```
It's extremely important to avoid duplicate `NodeModified` events.

It is more efficient to maintain a hash set of "already marked" nodes. You could reuse your more general algorithm (`GetDownstreamNodes`), which is currently on `EngineController`. 

So, I'd recommend moving your `GetDownstreamNodes` method to `NodeModel` and just do.

```csharp
var nodes = GetDownstreamNodes();
foreach ( var node in nodes ) {
   // mark modified
}
```

pboyer commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -602,6 +602,37 @@ public void NodeDeleted(NodeModel node)
+            foreach (var iNode in gathered)
+            {
+                syncDataManager.DeleteNodes(iNode.GUID);               
+            }
```
Nice, very clear.

pboyer commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -748,6 +773,16 @@ public virtual void OnNodeModified(bool forceExecute = false)
+
+            var handler = ToggleNodeFreeze;
+            if (handler != null) handler(this);
+        }
```
Why do we need this additional event? Why not use  the `PropertyChanged` event?

ramramps replied:
This is the same event as NodeModified. Just that this event causes the nodes to be deleted from AST.   Currently workspacemodel has no propertychangedevent for nodes. I dont want to introduce one there. Instead , went with event based model here.

pboyer replied:
What, NodeModel is doing 

```
RaisePropertyChanged("IsFrozen");
```

right before this call? Why not simply use that event and check the state of the property?

ramramps replied:
this is required to bubble up the event to UI. Otherwise, when you click freeze, the freeze will not be checked. Also, this I need this for setting up the node and connector display.

ramramps commented:
@pboyer  - PTAL. I have modified the code now.

ramramps commented:
ping @pboyer

pboyer commented on src/DynamoCore/Graph/Nodes/NodeModel.cs:
```diff
@@ -748,6 +792,16 @@ public virtual void OnNodeModified(bool forceExecute = false)
+        public event Action<NodeModel> ToggleNodeFreeze;
+        public virtual void OnToggleNodeFreeze()
+        {
+            if (!RaisesModificationEvents)
```
This is sort of a misuse of this boolean. It was only meant to silence `NodeModified` events. If you're generalizing it, you should update the documentation of this property appropriately.

ramramps replied:
Ah. I should remove this. I will do this now.

ramramps commented:
@pboyer - thanks for the comments. I removed the erroneous property changed. I should have done that in nodeviewmodel.  Updated the event name and the comments.
